transmart-core-db/grails-app/services/org/transmartproject/db/clinical/PatientSetService.groovy

```diff
@@ -152,27 +129,65 @@ class PatientSetService extends AbstractDataResourceService implements PatientSe
+        def queryInstance = new QtQueryInstance()
+        queryInstance.userId = user.username
+        queryInstance.groupId = Holders.config.getProperty('org.transmartproject.i2b2.group_id', String)
+        queryInstance.startDate = new Date()
+        queryInstance.statusTypeId = QueryStatus.PROCESSING.id
+        queryInstance.queryMaster = queryMaster
+        queryMaster.addToQueryInstances(queryInstance)
+        return queryInstance
+    }
+
+    protected QtQueryResultInstance createQueryResultInstance(QtQueryInstance queryInstance, String name) {
+        log.debug("Create a query result instance for ${queryInstance} query instance.")
+        def resultInstance = new QtQueryResultInstance()
+        resultInstance.statusTypeId = (short) QueryStatus.PROCESSING.id
+        resultInstance.startDate = new Date()
+        resultInstance.queryInstance = queryInstance
+        resultInstance.queryResultType = patientSetResultType
+        resultInstance.description = name
+        return resultInstance
+    }
+
     /**
-     * Find a query result based on a constraint.
+     * Find a query result based on a constraint and user in the cache.
      *
      * @param user the creator of the query result.
      * @param constraint the constraint used in the lookup.
-     * @return the query result if it exists; null otherwise.
+     * @return the query result if it exists in the cache and the database; null otherwise.
      */
-    QueryResult findQueryResultByConstraint(User user,
-                                            Constraint constraint) {
-        def criteria = DetachedCriteria.forClass(QtQueryResultInstance.class, 'qri')
-                .createCriteria('qri.queryInstance', 'qi')
-                .createCriteria('qi.queryMaster', 'qm')
-                .add(Restrictions.eq('qri.queryResultType', patientSetResultType))
-                .add(Restrictions.eq('qri.deleteFlag', 'N'))
-                .add(Restrictions.eq('qri.statusTypeId', (short)QueryStatus.FINISHED.id))
-                .add(Restrictions.eq('qi.userId', user.username))
-                .add(Restrictions.eq('qi.deleteFlag', 'N'))
-                .add(Restrictions.eq('qm.requestConstraints', constraint.toJson()))
-                .add(Restrictions.eq('qm.deleteFlag', 'N'))
-                .addOrder(Order.desc('qri.endDate'))
-        (QueryResult)getFirst(criteria)
+    QueryResult findFinishedQueryResultInCacheBy(User user,
+                                                 Constraint constraint) {
+        Long queryResultId = getQueryResultIdFromCache(user, constraint)
+        if (log.debugEnabled) {
+            log.debug("Get query result ${queryResultId} for ${user} and ${constraint}")
+        }
+        if (queryResultId) {
+            QtQueryResultInstance queryResult = QtQueryResultInstance.get(queryResultId)
+            boolean usable = queryResult && queryResult.status == QueryStatus.FINISHED && queryResult.deleteFlag == 'N'
```

Comment:
If you reverse the order, the massive indenting is not required and also a separate function to check this is not needed. E.g.,:
```groovy
if (!queryResultId) {
  return null
}
...
if (!queryResult || queryResult.status != FINISHED) {
  return null
}
```